File: mobile/android/app/src/main/java/com/tigerwallet/app/trading/CopyTradingService.java
package com.tigerwallet.app.trading;

/**
 * Copy Trading Service - Android Implementation
 *
 * WARNING: This orphan service layer previously fabricated a hardcoded list of
 * fake traders (invented addresses 0x1234.../0xabcd..., invented win rates and
 * follower counts). Those methods now throw UnsupportedOperationException.
 * Real copy-trading data is served by the canonical Go copy_trading_service
 * (:8006, GET /api/v1/copytrading/traders) and consumed by the canonical
 * Android app (mobile_apps/android_app). The pure data classes below are
 * retained for callers that construct Trader/CopyPosition from real backend
 * responses.
 */

import java.util.List;

public class CopyTradingService {

    public static class Trader {
        public String id;
        public String address;
        public String username;
        public String avatar;
        public double winRate;
        public double totalPnl;
        public int followers;
        public int copyCount;
        public String tradingPair;
        public double monthlyPnl;
        public double weeklyPnl;
        public double dailyPnl;
        public String riskLevel;
        public boolean isFollowing;
        public boolean isVerified;
    }

    public static class CopyPosition {
        public String id;
        public String traderId;
        public String traderName;
        public String userId;
        public String symbol;
        public String side;
        public double size;
        public double entryPrice;
        public double currentPrice;
        public double pnl;
        public double pnlPercent;
        public long openTime;
        public String status;
    }

    public CopyTradingService() {}

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer;
     *                                       use the canonical Android app.
     */
    public List<Trader> getTopTraders(int limit) {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured. Use the canonical " +
            "copy_trading_service (/api/v1/copytrading/traders) via the " +
            "mobile_apps/android_app client.");
    }

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer.
     */
    public List<Trader> getAllTraders() {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured in this orphan layer.");
    }

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer.
     */
    public Trader getTrader(String traderId) {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured in this orphan layer.");
    }

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer.
     */
    public void followTrader(String traderId) {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured in this orphan layer.");
    }

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer.
     */
    public CopyPosition copyTrade(String userId, String traderId, String symbol, String side, double amount) {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured in this orphan layer.");
    }

    /**
     * @throws UnsupportedOperationException no real copy-trading backend is
     *                                       configured in this orphan layer.
     */
    public List<Trader> searchTraders(String query) {
        throw new UnsupportedOperationException(
            "No real copy-trading backend configured in this orphan layer.");
    }
}

File: mobile_apps/ios_legacy/TigerWallet/Services/Trading/CopyTradingService.swift
//
//  CopyTradingService.swift
//  TigerWallet
//
//  Copy Trading Service - iOS Implementation
//

import Foundation

struct Trader {
    var id: String
    var address: String
    var username: String
    var avatar: String
    var winRate: Double
    var totalPnl: Double
    var followers: Int
    var copyCount: Int
    var tradingPair: String
    var monthlyPnl: Double
    var weeklyPnl: Double
    var dailyPnl: Double
    var riskLevel: String
    var isFollowing: Bool
    var isVerified: Bool
}

struct CopyPosition {
    var id: String
    var traderId: String
    var traderName: String
    var userId: String
    var symbol: String
    var side: String
    var size: Double
    var entryPrice: Double
    var currentPrice: Double
    var pnl: Double
    var pnlPercent: Double
    var openTime: Date
    var status: String
}

class CopyTradingService {
    
    private var traders: [Trader] = []
    
    init() {
        // Traders are NOT seeded with hardcoded mock data. The real trader
        // leaderboard is fetched from the backend copy-trading service
        // (loadTradersFromBackend) — never fabricated addresses/PnL.
        loadTradersFromBackend()
    }

    private func loadTradersFromBackend() {
        // Real trader data is fetched from the backend; until then the list
        // is empty (honest) rather than seeded with fake traders.
        traders = []
    }

    func getTopTraders(limit: Int = 10) -> [Trader] {
        return Array(traders.prefix(limit))
    }
    
    func getAllTraders() -> [Trader] {
        return traders
    }
    
    func getTrader(traderId: String) -> Trader? {
        return traders.first { $0.id == traderId }
    }
    
    func followTrader(traderId: String) {
        if let index = traders.firstIndex(where: { $0.id == traderId }) {
            traders[index].isFollowing.toggle()
            traders[index].followers += traders[index].isFollowing ? 1 : -1
        }
    }
    
    func copyTrade(userId: String, traderId: String, symbol: String, side: String, amount: Double) -> CopyPosition {
        let trader = getTrader(traderId: traderId)
        
        return CopyPosition(
            id: "copy_\(Int(Date().timeIntervalSince1970 * 1000))",
            traderId: traderId,
            traderName: trader?.username ?? "Unknown",
            userId: userId,
            symbol: symbol,
            side: side,
            size: amount,
            entryPrice: 43250.0,
            currentPrice: 43250.0,
            pnl: 0,
            pnlPercent: 0,
            openTime: Date(),
            status: "OPEN"
        )
    }
    
    func searchTraders(query: String) -> [Trader] {
        return traders.filter {
            $0.username.lowercased().contains(query.lowercased()) ||
            $0.address.lowercased().contains(query.lowercased())
        }
    }
}
